sdk/program/src/sysvar/slot_hashes.rs

```diff
@@ -97,30 +100,50 @@ impl SlotHashesSysvar {
     /// Get the position of an entry in the sysvar by its key.
     /// Returns `None` if the key is not found.
     pub fn position(slot: &Slot) -> Result<Option<usize>, ProgramError> {
-        get_pod_slot_hashes().map(|pod_hashes| {
+        Self::get_pod_slot_hashes().map(|pod_hashes| {
             pod_hashes
                 .binary_search_by(|PodSlotHash { slot: this, .. }| slot.cmp(this))
                 .ok()
         })
     }
-}
 
-fn get_pod_slot_hashes() -> Result<Vec<PodSlotHash>, ProgramError> {
-    let mut pod_hashes = vec![PodSlotHash::default(); MAX_ENTRIES];
-    {
-        let data = bytemuck::try_cast_slice_mut::<PodSlotHash, u8>(&mut pod_hashes)
-            .map_err(|_| ProgramError::InvalidAccountData)?;
-
-        // Ensure the created buffer is aligned to 8.
-        if data.as_ptr().align_offset(8) != 0 {
-            return Err(ProgramError::InvalidAccountData);
+    /// Return the slot hashes sysvar as a vector of `PodSlotHash`.
+    pub fn get_pod_slot_hashes() -> Result<Vec<PodSlotHash>, ProgramError> {
+        // First fetch all the sysvar data.
+        let sysvar_len = SlotHashes::size_of();
+        let mut data = vec![0; sysvar_len];
+        get_sysvar(
+            &mut data,
+            &SlotHashes::id(),
+            /* offset */ 0,
+            /* length */ sysvar_len as u64,
+        )?;
+
+        // Read the sysvar's vector length (u64).
+        let slot_hash_count = data
+            .get(..U64_SIZE)
+            .and_then(|bytes| bytes.try_into().ok())
+            .map(usize::from_le_bytes)
+            .ok_or(ProgramError::InvalidAccountData)?;
+
+        // If the vector length is 0, return an empty vector.
+        if slot_hash_count == 0 {
+            return Ok(vec![]);
         }
 
-        let offset = 8; // Vector length as `u64`.
-        let length = (SlotHashes::size_of() as u64).saturating_sub(offset);
-        get_sysvar(data, &SlotHashes::id(), offset, length)?;
+        // From the vector length, determine the expected length of the data.
+        let length = slot_hash_count
+            .checked_mul(std::mem::size_of::<SlotHash>())
+            .ok_or(ProgramError::ArithmeticOverflow)?;
+        let start = U64_SIZE;
+        let end = start.saturating_add(length);
+
+        // Finally, convert to a vector of `PodSlotHash`.
+        data.get(start..end)
+            .and_then(|data| bytemuck::try_cast_slice(data).ok())
+            .map(|pod_hashes| pod_hashes.to_vec())
```

Comment:
I might be missing something, but I believe `to_vec()` will *redo* the huge allocation -- we should instead find a way to reuse the existing buffer that was already fetched to avoid blowing up the heap.

I initially thought that `SlotHashesSysvar` would hold onto the underlying `Vec<u8>` and that it would also expose a way to get the underlying `PodSlotHash` slice using something like `as_slice()`, which would do the bytemuck cast.

Otherwise, we'll need to do some unsafe stuff here to lop off the first 8 bytes and return the rest as `Vec<PodSlotHash>` without doing another allocation. Although, looking at the Vec interface, I think it'll be pretty gross do that unfortunately.

Any better ideas for this?

Reply:
How about doing a slice from raw parts for the initial mutable buffer, since we know it's a fixed size, and then that call `.to_vec()` should do just one heap allocation, since the original slice will be stack?

Reply:
Sorry, I might be misunderstanding -- we can't have a 20kb variable on the stack until dynamic stack frames are added, so what would the "initial mutable buffer" be?

Reply:
Ahh darn, yeah that probably won't work then. I think I see what you're getting at in your first message. I've added a commit that hangs onto the initial buffer, and then the commit right after uses `as_slice`. Let me know if this is what you had in mind.